libass/ass_filesystem.c

```diff
@@ -0,0 +1,410 @@
+    }
+    free(wpath);
+    if (dir->handle == INVALID_HANDLE_VALUE)
+        return false;
+
+    size = NAME_BUF_SIZE + 2;
+    size_t wlen = wcslen(data.cFileName);
+    if (!check_add_size(&size, path.len) || !check_add_size_wtf16to8(&size, wlen) ||
+            !(dir->path = malloc(size))) {
+        FindClose(dir->handle);
+        return false;
+    }
+    dir->max_path = size;
+    memcpy(dir->path, path.str, path.len);
+    if (add_separator)
+        dir->path[path.len++] = '\\';
+    *convert_wtf16to8(dir->path + path.len, data.cFileName, wlen) = '\0';
+    dir->prefix = path.len;
+    return true;
+}
+
+bool ass_open_dir(ASS_Dir *dir, const char *path)
+{
+    dir->handle = INVALID_HANDLE_VALUE;
+    dir->path = NULL;
+    dir->name = NULL;
+
+    size_t len = strlen(path);
+    if (open_dir_wtf8(dir, (ASS_StringView) { path, len }))
+        return true;
+
+    if (len > INT_MAX)
+        return false;
+
+    UINT cp = CP_ACP;
+#if ASS_WINAPI_DESKTOP
+    if (!AreFileApisANSI())
+        cp = CP_OEMCP;
+#endif
+    size_t wlen = MultiByteToWideChar(cp, 0, path, len, NULL, 0);
+    if (wlen > (SIZE_MAX - sizeof(dir_tail)) / sizeof(WCHAR))
+        return false;
+    WCHAR *wpath = malloc(wlen * sizeof(WCHAR) + sizeof(dir_tail));
+    if (!wpath)
+        return false;
+    MultiByteToWideChar(cp, 0, path, len, wpath, wlen);
+    bool add_separator = append_tail(wpath, wlen);
+
+    WIN32_FIND_DATAW data;
+    dir->handle = FindFirstFileExW(wpath, FindExInfoBasic, &data, FindExSearchNameMatch, NULL, 0);
+    if (dir->handle == INVALID_HANDLE_VALUE) {
+        free(wpath);
+        return false;
+    }
+    size_t size = NAME_BUF_SIZE + 2, wlen1 = wcslen(data.cFileName);
+    if (!check_add_size_wtf16to8(&size, wlen) || !check_add_size_wtf16to8(&size, wlen1) ||
+            !(dir->path = malloc(size))) {
+        FindClose(dir->handle);
+        return false;
+    }
+    dir->max_path = size;
+    char *ptr = convert_wtf16to8(dir->path, wpath, wlen);
+    if (add_separator)
+        *ptr++ = '\\';
+    convert_wtf16to8(ptr, data.cFileName, wlen1 + 1);
+    dir->prefix = ptr - dir->path;
+    return true;
+}
+
+const char *ass_read_dir(ASS_Dir *dir)
+{
+    if (!dir->name)  // first invocation
+        return dir->name = dir->path + dir->prefix;
+
+    WIN32_FIND_DATAW data;
+    while (FindNextFileW(dir->handle, &data)) {
+        size_t size = dir->prefix + 1, wlen = wcslen(data.cFileName);
+        if (!check_add_size_wtf16to8(&size, wlen) || !alloc_path(dir, size))
+            continue;
+        convert_wtf16to8(dir->path + dir->prefix, data.cFileName, wlen + 1);
+        return dir->name = dir->path + dir->prefix;
+    }
+    return NULL;
+}
+
+const char *ass_current_file_path(ASS_Dir *dir)
+{
+    return dir->path;
+}
+
+void ass_close_dir(ASS_Dir *dir)
+{
+    if (dir->handle != INVALID_HANDLE_VALUE)
+        FindClose(dir->handle);
+    free(dir->path);
+    dir->handle = INVALID_HANDLE_VALUE;
+    dir->path = NULL;
+}
```

Comment:
Since in this variant `dir->name` points into `dir->path`s memory it definitely becomes a dangling pointer and should also be set to `NULL`. Also it being `NULL`is used to detect the first readdir call.

Reply:
Done.